### PR TITLE
Fix count issue when deep filtering

### LINKED ISSUE #7631: Count Metadata doesn't take filtering into account
Hello.

I have this situation, where `{entity}/count` REST endpoint with deep filtering and using '_or' clause is not working properly. Count value is even bigger than number of items that are stored in my db (of that type) at time of request.

So we have these two tables: Dogs and Colors. Every dog can have multiple colors assigned using following relation:

Screenshot 2020-08-27 at 19 13 11

Please look at following picture to see, what query params are comming from our Frontend to strapi endpoint (endpoint is overriden only for purposes of this test. We normally use pre-generated endpoint):

Screenshot 2020-08-27 at 19 26 00

As you can see, I have 11 dogs in my development DB, but count endpoint has provided **15** as result.

**Cause of the issue is the most probably a fact, that some of dogs have multiple colors assigned. When I use only 1 color per dog, thne the filtering problem dissapears.**

Does anyone know, what could be the reason for this behaviour?

Thank you so much and have a nice day!

**System**

- Strapi version: 3.1.4
- Database: tried on SQLite and PostgreSQL
- Operating system: macOS Catalina

### PR DESCRIPTION
**Pull request by Bassel17**
### What does it do?

Changes the count implementation in the query builder, so it uses `countDistinct()` based on the column id

### Why is it needed?

When counting with deep filters columns were counted multiple times based on their relation

### How to test it?

Create a content type with a relation (one-to-many or many-to-many) (you can take a look at the issue linked below for more details) add multiple elements to the relation field and then try to get the content type while filtering on that field, the total in the meta should be correct.

### Related issue(s)/PR(s)

fix #7631

### CODE REVIEW

**Review by innerdvations — APPROVED** [review]

File: packages/core/database/lib/query/query-builder.js [hunk 1]
```diff
@@ -301,10 +301,8 @@ const createQueryBuilder = (uid, db) => {
           break;
         }
         case 'count': {
-          const dbColumnName =
-            state.count === '*' ? '*' : this.aliasColumn(helpers.toColumnName(meta, state.count));
-
-          qb.count({ count: dbColumnName });
+          const dbColumnName = this.aliasColumn(helpers.toColumnName(meta, 'id'));
```
Comment (alexandrebodin) [inline packages/core/database/lib/query/query-builder.js:304]: This removes a feature for counting on a specific column. This is a breaking change.

**Review by alexandrebodin — CHANGES_REQUESTED** [review]

File: packages/core/database/lib/query/query-builder.js (hunk 1 above)
Comment (Bassel17) [reply packages/core/database/lib/query/query-builder.js:304]: @alexandrebodin ok, how do we exactly count by a specific column?

**Comment by Kyle772** [comment]
Thank you for pursuing this @Bassel17 and @alexandrebodin !

File: packages/core/database/lib/query/query-builder.js (hunk 1 above)
Comment (derrickmehaffy) [reply packages/core/database/lib/query/query-builder.js:304]: I was not aware that counting on a specific column existed either, sorting yes obviously but counting I believe should be based on the ID.

Unless this is an undocumented feature?

Comment (Kyle772) [reply packages/core/database/lib/query/query-builder.js:304]: Pardon the ping but I ran into this issue today with a total count that includes a deeply nested self reference which might actually be caused by this column count feature mentioned here. The issue may not be with the actual counting logic but rather the column counting logic more specifically 

schema is roughly: 
```
product: {
    make: { // relation
        models // relation
    },
    model: { // relation
        years, make // two relations
    },
    year: { // relation
        make, model // two relations
    },
}
```

In my scenario it seems like the total count should be 1 but is actually 12 (max for the request)  I assume this is coming from size * count but pageCount should not be 12. The data returned is correct but the pagination count appears to be going multiple layers deep looking for matches (and possibly populating all references all the way down further than my request is intending).

pagination response from request looking for make_name and model_name matches on one specific slug each 
{
	"pagination": {
		"page": 1,
		"pageSize": 12,
		"pageCount": 12,
		"total": 144
	}
}

Comment (alexandrebodin) [reply packages/core/database/lib/query/query-builder.js:304]: Counting on a column is a SQL feature. It excludes null values so you can count the real number of rows that actually have a value for a column. 

It is also useful when using distinct counts as it avoids having to use group by statements. 

@Bassel17 you can update the code slightly to still accept the `state.count` if defined and use id if not so we can keep that feature.

Comment (Bassel17) [reply packages/core/database/lib/query/query-builder.js]: @alexandrebodin yes got it, I pushed a fix for it

File: packages/core/database/lib/query/query-builder.js [hunk 2]
```diff
@@ -302,9 +302,11 @@ const createQueryBuilder = (uid, db) => {
         }
         case 'count': {
           const dbColumnName =
-            state.count === '*' ? '*' : this.aliasColumn(helpers.toColumnName(meta, state.count));
+            state.count === '*'
+              ? this.aliasColumn(helpers.toColumnName(meta, 'id'))
+              : this.aliasColumn(helpers.toColumnName(meta, state.count));
 
-          qb.count({ count: dbColumnName });
+          qb.countDistinct({ count: dbColumnName });
```
Comment (alexandrebodin) [inline packages/core/database/lib/query/query-builder.js:309]: Now we need to do a distinct count only when necessary has it has a big impact on performance and if we don't do deep filtering it is not needed.

I think you can use the same if logic as in the select 'shouldUseDistinct' to optimize this

**Review by alexandrebodin — CHANGES_REQUESTED** [review]

File: packages/core/database/lib/query/query-builder.js (hunk 2 above)
Comment (Bassel17) [reply packages/core/database/lib/query/query-builder.js]: @alexandrebodin I eidted the `shouldUseDistinct` a bit and used to fit my case

File: packages/core/database/lib/query/query-builder.js [hunk 3]
```diff
@@ -302,9 +302,15 @@ const createQueryBuilder = (uid, db) => {
...
         case 'count': {
           const dbColumnName =
-            state.count === '*' ? '*' : this.aliasColumn(helpers.toColumnName(meta, state.count));
-
-          qb.count({ count: dbColumnName });
+            state.count === '*'
+              ? this.aliasColumn(helpers.toColumnName(meta, 'id'))
+              : this.aliasColumn(helpers.toColumnName(meta, state.count));
+
+          if (this.shouldUseDistinct(2)) {
```
Comment (alexandrebodin) [inline packages/core/database/lib/query/query-builder.js:309]: why do you need to have at least 2 joins?

File: packages/core/database/lib/query/query-builder.js [hunk 4]
```diff
@@ -302,9 +302,15 @@ const createQueryBuilder = (uid, db) => {
         }
         case 'count': {
           const dbColumnName =
-            state.count === '*' ? '*' : this.aliasColumn(helpers.toColumnName(meta, state.count));
-
-          qb.count({ count: dbColumnName });
+            state.count === '*'
```
Comment (alexandrebodin) [inline packages/core/database/lib/query/query-builder.js:305]: this check seems a bit weird. we shouldn't transform * to id if the user wants to use *
I think you should change the default count to be 'id' (Line 74) and simplify the code here to always use state.count

**Review by alexandrebodin — CHANGES_REQUESTED** [review]

File: packages/core/database/lib/query/query-builder.js (hunk 3 above)
Comment (Bassel17) [reply packages/core/database/lib/query/query-builder.js:309]: The counting issue is really only present when we have multiple deep filters, and so considering as you said before that count distinct has a big impact on performance I wanted to make sure that we only use it when necessary essentially

File: packages/core/database/lib/query/query-builder.js (hunk 4 above)
Comment (Bassel17) [reply packages/core/database/lib/query/query-builder.js:305]: yeah that does make sense makes sense

File: packages/core/database/lib/query/query-builder.js (hunk 3 above)
Comment (alexandrebodin) [reply packages/core/database/lib/query/query-builder.js:309]: as soon as you make a join you might have duplicated rows 🤔 deep filtering can imply only one join and still create problems so I would keep the old distinct here :)

Comment (Bassel17) [reply packages/core/database/lib/query/query-builder.js:309]: ok sounds good @alexandrebodin

**Review by alexandrebodin — APPROVED** [review]
LGTM

**Pull request merged**